### PR TITLE
DSD-555: Structured Content

### PR DESCRIPTION
**Pull request by EdwinGuzman**
Fixes JIRA ticket [DSD-555](https://jira.nypl.org/browse/DSD-555)

## This PR does the following:

- Adds the `StructuredContent` component. Still a work-in-progress.

@bigfishdesign13  can you check the progress so far before I proceed? I'm more worried about the `Image`'s aspect ratio in whenever the `imageAspectRatio` prop is _not_ set to `ImageRatios.Original`. The following screenshots use `ImageRatios.Square` (the default) and you can see it takes the width of the caption (and credit), otherwise, it displays in its desired size:

- with a longish caption/credit
Screen Shot 2022-01-03 at 4 32 17 PM

- with a short caption/credit
Screen Shot 2022-01-03 at 4 32 26 PM

- no caption/credit
Screen Shot 2022-01-03 at 4 33 06 PM

## How has this been tested?

No tests yet!

## Accessibility concerns or updates

- Should we use the `section` and `article` HTML elements?

### Checklist:

- [ ] I have updated the Storybook documentation accordingly.
- [ ] I have added relevant accessibility documentation for this pull request.
- [ ] All new and existing tests passed.

### Front End Review:

- [ ] View [the example in Storybook]()

### CODE REVIEW

File: src/theme/components/structuredContent.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  baseStyle: ({ imagePosition }) => {
+    const styles = {};
+    if (imagePosition === "center") {
+      (styles as any).maxWidth = "100%";
+    }
+    return {
+      image: {
+        float: imagePosition !== "center" ? imagePosition : undefined,
+        marginLeft:
+          imagePosition === "center"
```
Comment (EdwinGuzman) [inline src/theme/components/structuredContent.ts]: This isn't complicated but hard to read.

File: src/components/StructuredContent/StructuredContent.stories.mdx [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+              pharetra.
+            </i>
+          </p>
+        </>
+      }
+    />
+  </Story>
+</Canvas>
+
+## Examples
```
Comment (EdwinGuzman) [inline src/components/StructuredContent/StructuredContent.stories.mdx:249]: Tried to make this resemble the shape of the examples in Figma but I can update the text to better match later on.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

File: src/theme/components/structuredContent.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+            ? null
+            : imagePosition === "left"
+            ? "0"
+            : "s",
```
Comment (bigfishdesign13) [inline src/theme/components/structuredContent.ts]: This should be `m` (24px).

File: src/theme/components/structuredContent.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+const StructuredContent = {
+  parts: ["image"],
+  baseStyle: ({ imagePosition }) => {
+    const styles = {};
+    if (imagePosition === "center") {
+      (styles as any).maxWidth = "100%";
+    }
+    return {
+      image: {
```
Comment (bigfishdesign13) [inline src/theme/components/structuredContent.ts:34]: Add `marginBottom: m`.  Currently, the image and captions can be pressed right up against text that runs below the image.

File: src/components/StructuredContent/StructuredContent.tsx [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+import generateUUID from "../../helpers/generateUUID";
+
+export enum StructuredContentImagePosition {
+  Left = "left",
+  Right = "right",
+  Center = "center",
+}
+
+export interface StructuredContentProps {
+  calloutContent?: string | JSX.Element;
```
Comment (bigfishdesign13) [inline src/components/StructuredContent/StructuredContent.tsx]: I am reconsidering this prop type assignment.  I think we can drop `JSX.Element` and just use `string`.  I would like the formatting of the `Callout` heading to always be consitent.

File: src/components/StructuredContent/StructuredContent.tsx [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    className,
+    headingText,
+    id = generateUUID(),
+    imageAlt = "",
+    imageAspectRatio = ImageRatios.Original,
+    imageCaption,
+    imageComponent,
+    imageCredit,
+    imagePosition = StructuredContentImagePosition.Left,
+    imageSize = ImageSizes.Medium,
```
Comment (bigfishdesign13) [inline src/components/StructuredContent/StructuredContent.tsx:96]: This is important and I will explain in my review comment.

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]

**Comment by bigfishdesign13** [comment]
@EdwinGuzman Sorry, I submitted my review before adding my comment.

Overall, this looks pretty good.  I left a few code comments, but I also wanted to talk about the image size issue that you noted.  Similar to `Card`, when `imageSize` is set to `default`, this component needs to set a specific width value and not simply use `100%`.  `Card` sets the image to the `medium` size when `imageSize` is set to default.  `StructuredContent` should do the same.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

**Comment by EdwinGuzman** [comment]
@bigfishdesign13 This is now ready to be fully reviewed. The image is now rendering correctly, on either the left or right side, when the size and aspect ratio are updated.

**Comment by EdwinGuzman** [comment]
I mentioned this in the spec doc but I changed the callout content heading from the initial level 4 to level 3 with the display size as "callout". Otherwise, we get this accessibility warning:
Screen Shot 2022-01-05 at 10 07 55 AM

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

File: src/components/StructuredContent/StructuredContent.tsx [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+import Heading from "../Heading/Heading";
+import { HeadingDisplaySizes, HeadingLevels } from "../Heading/HeadingTypes";
+import Image, { ImageProps } from "../Image/Image";
+import { ImageRatios, ImageSizes } from "../Image/ImageTypes";
+import { StructuredContentImagePosition } from "./StructuredContentTypes";
+import generateUUID from "../../helpers/generateUUID";
+
+export interface StructuredContentProps {
+  calloutContent?: string;
```
Comment (bigfishdesign13) [inline src/components/StructuredContent/StructuredContent.tsx]: Now that this prop is only a string, what do you think about changing the name to `calloutText` in order to follow the same pattern as `headingText`?

File: src/components/StructuredContent/StructuredContent.tsx [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+  id?: string;
+  imageAlt?: string;
+  imageAspectRatio?: ImageRatios;
+  imageCaption?: string;
+  imageComponent?: JSX.Element;
+  imageCredit?: string;
+  imagePosition?: StructuredContentImagePosition;
+  imageSize?: ImageSizes;
+  imageSrc?: string;
+  textContent: string | JSX.Element;
```
Comment (bigfishdesign13) [inline src/components/StructuredContent/StructuredContent.tsx]: If we change `calloutContent` to `calloutText`, I feel `textContent` is not an appropriate name for this prop.  What about `bodyContent` or `mainContent`?

File: src/components/StructuredContent/StructuredContent.tsx [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+  });
+  const finalTextContent =
+    typeof textContent === "string" ? (
+      <div dangerouslySetInnerHTML={{ __html: textContent }} />
+    ) : (
+      <Box>{textContent}</Box>
+    );
+
+  if (hasImage && !imageAlt) {
+    console.warn(
```
Comment (bigfishdesign13) [inline src/components/StructuredContent/StructuredContent.tsx:115]: Can you also add a warning when `imagePosition` is `center` and `imageSize` is set to anything other than `default`?

File: src/theme/components/structuredContent.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        marginRight:
+          imagePosition === "center"
+            ? null
+            : imagePosition === "left"
+            ? "m"
+            : "0",
+        width: imagePosition !== "center" ? "100%" : undefined,
+        ...styles,
+      },
+      image: {
```
Comment (bigfishdesign13) [inline src/theme/components/structuredContent.ts:34]: The `image` also needs `marginBottom: "m"` when the `imageContainer` is not used (i.e. no caption and no credit).  See the last example on the story page.

File: src/components/StructuredContent/StructuredContent.stories.mdx [hunk 11]
```diff
@@ -0,0 +1,377 @@
...
+      imageSize={ImageSizes.Medium}
+      imageSrc="https://placeimg.com/400/300/animals"
+      textContent={
+        <>
+          <p>
+            Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
+            eiusmod tempor incididunt ut labore et dolore magna aliqua.
+            Pellentesque massa placerat duis ultricies lacus sed turpis
+            tincidunt.
+            <a href="#">Porttitor eget dolor</a> morbi non arcu risus quis varius.
```
Comment (bigfishdesign13) [inline src/components/StructuredContent/StructuredContent.stories.mdx]: Bring this line up with the end of the previous sentence.  When it's broken like this, the space between the `.` and the link text is lost.

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
The image is working well now.  There are just a few other things that need to be cleaned up.

File: src/components/StructuredContent/StructuredContent.tsx (hunk 7 above)
Comment (EdwinGuzman) [reply src/components/StructuredContent/StructuredContent.tsx]: Yeah, that better matches up with `headingText`.

File: src/components/StructuredContent/StructuredContent.tsx (hunk 8 above)
Comment (EdwinGuzman) [reply src/components/StructuredContent/StructuredContent.tsx]: let's go with `bodyContent`.

**Comment by EdwinGuzman** [comment]
@bigfishdesign13 updated.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

File: src/components/StructuredContent/StructuredContent.tsx [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+   * Defaults to `StructuredContentImagePosition.Left`. */
+  imagePosition?: StructuredContentImagePosition;
+  /** Optional value to control the size of the internal `Image` component.
+   * Defaults to `ImageSizes.Medium`. */
+  imageSize?: ImageSizes;
+  /** Optional value that contains the path to an image. If omitted, the internal
+   * DS `Image` component will not render. */
+  imageSrc?: string;
+  /** Required value to set the text for the body content. */
+  bodyContent: string | JSX.Element;
```
Comment (bigfishdesign13) [inline src/components/StructuredContent/StructuredContent.tsx:42]: Small nit: alphabetize based on updated prop name

File: src/theme/components/structuredContent.ts (hunk 10 above)
Comment (bigfishdesign13) [reply src/theme/components/structuredContent.ts:34]: This needs some more work.  I'll setup a call so we can quickly go over this together.

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
The logic behind the image margins needs some additional work.  I will setup a call so we can talk about it.

**Comment by EdwinGuzman** [comment]
Updated.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

File: src/theme/components/structuredContent.ts [hunk 13]
```diff
@@ -1,34 +1,47 @@
...
-        marginBottom: noImageWrapper ? "m" : null,
+        marginLeft:
+          imagePosition === "right" && imageAspectRatio === "original"
+            ? "m"
+            : undefined,
+        marginRight:
+          imagePosition === "left" && imageAspectRatio === "original"
+            ? "m"
+            : undefined,
+        marginBottom: "var(--nypl-space-m) !important",
```
Comment (bigfishdesign13) [inline src/theme/components/structuredContent.ts]: This style needs some logic applied.  The way this is written now, the bottom margin will always be applied to the image element and there is a gap between the image and the image caption.  

However, I see above the style was originally written as:
```marginBottom: noImageWrapper ? "m" : null,```

That seems like it would have been correct.  Why the change?

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
Some of the changes work well, but the bottom margin on the image is still being applied when it should not.

image

**Comment by EdwinGuzman** [comment]
> the bottom margin on the image is still being applied when it should not.

Okay, I got it reversed and thought it _should_ have the bottom margin. This also explains why I changed the logic for the bottom margin for the image in its style file.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

File: src/components/StructuredContent/StructuredContent.tsx [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+      "StructuredContent: `imageAlt` prop is required when using an image."
+    );
+  }
+
+  if (
+    imagePosition === StructuredContentImagePosition.Center &&
+    imageSize !== ImageSizes.Default
+  ) {
+    console.warn(
+      `StructuredContent: image size "${imageSize}" is not supported when using the center image position.`
```
Comment (bigfishdesign13) [inline src/components/StructuredContent/StructuredContent.tsx]: The new "additional" styles functionality that was added to `Image` seems to have provided more control over  `Image` in `StructuredContent` and the image is now resizing properly for the center position.  Which I like.  Because of that, this warning message is no longer relevant and can be removed.

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
This is working very nicely now.  Becuase of the added control of the `Image` styles, one of the warning messages can now be removed.

**Comment by EdwinGuzman** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr827-2iyp0sq7qauranirqe2bkusz7k72oja8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/61d36bed7499db65228b92ef

**Review by bigfishdesign13 — APPROVED** [review]
Very nice!
